app/controllers/search_controller.rb

```diff
@@ -0,0 +1,28 @@
+class SearchController < ApplicationController
+  skip_authorization_check
+  
+  respond_to :html, :json, :js
+
+  def index
+    @tagged_products = Product.includes(:reviews).tagged_with(params[:query])
+    @product_results = @tagged_products + PgSearch.multisearch(params[:query]).includes(:searchable).where(searchable_type: 'Product').map(&:searchable)
+    @product_results = Kaminari.paginate_array(@product_results).page(params[:product]).per(4)
+
+    @article_results = PgSearch.multisearch(params[:query]).includes(:searchable).where(searchable_type: 'Article').map(&:searchable)
+    @article_results = Kaminari.paginate_array(@article_results).page(params[:article]).per(4)
+
+    @reference_results = PgSearch.multisearch(params[:query]).includes(:searchable).where(searchable_type: 'Reference').map(&:searchable)
+    @reference_results = Kaminari.paginate_array(@reference_results).page(params[:resource]).per(4)
+
+    @tag_results = PgSearch.multisearch(params[:query]).includes(:searchable).where(searchable_type: 'Tag').map(&:searchable)
+    @tag_results = Kaminari.paginate_array(@tag_results).page(params[:tag]).per(4)
```

Comment:
I'd rewrite these to not be (unnecessarily) declaring an instance variable and then overwriting it immediately after; maybe something like:
```
searchable_results = PgSearch.multisearch(params[:query]).includes(:searchable)

@tagged_products = Product.includes(:reviews).tagged_with(params[:query])
products = @tagged_products + searchable_results.where(searchable_type: 'Product').pluck(:searchable)
@product_results = Kaminari.paginate_array(products).page(params[:product]).per(4)

articles = searchable_results.where(searchable_type: 'Article').pluck(:searchable)
@article_results = Kaminari.paginate_array(articles).page(params[:article]).per(4)

refereneces = searchable_results.where(searchable_type: 'Reference').pluck(:searchable)
@reference_results = Kaminari.paginate_array(refereneces).page(params[:resource]).per(4)

tags = searchable_results.where(searchable_type: 'Tag').pluck(:searchable)
@tag_results = Kaminari.paginate_array(tags).page(params[:tag]).per(4)
```